File: github_code/merge-sort/java/84.java
class MergeSort {

  void mergeSort(int[] arr, int l, int r) {
    if (l < r) {
      int mid = (l + r) / 2;
      mergeSort(arr, l, mid);
      mergeSort(arr, mid + 1, r);
      merge(arr, l, r, mid);
    }
  }

  void merge(int[] arr, int l, int r, int mid) {
    int n1 = mid - l + 1;
    int n2 = r - mid;

    int L[] = new int [n1];
    int R[] = new int [n2];

    for (int i = 0; i < n1; i++)
      L[i] = arr[l + i];
    for (int j = 0; j < n2; j++)
      R[j] = arr[mid + j + 1];

    int i = 0;
    int j = 0;
    int k = l;

    while (i < n1 && j < n2) {
      if (L[i] <= R[j]) {
        arr[k] = L[i];
        i++;
      } else {
        arr[k] = R[j];
        j++;
      }
      k++;
    }

    while (i < n1) {
      arr[k] = L[i];
      i++;
      k++;
    }

    while (j < n2) {
      arr[k] = R[j];
      j++;
      k++;
    }
  }

  void printArray(int[] arr) {
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }
    System.out.println();
  }

  public static void main(String[] args) {
    MergeSort sort = new MergeSort();
    int[] arr = {37, 27, 43, 3, 9, 82 , 10};
    sort.mergeSort(arr, 0, arr.length - 1);
    sort.printArray(arr);
  }
}

// mergeSort(arr, 0, 6);  l = 0, r = 6, mid = 3
// mergeSort(arr, 0, 3);  l = 0, r = 3, mid = 1
// mergeSort(arr, 0, 1);  l = 0, r = 1, mid = 0
// mergeSort(arr, 0, 0);  l = 0, r = 0, mid = 0
//
// mergeSort(arr, 0, 6);  l = 0, r = 6, mid = 3
// mergeSort(arr, 0, 3);  l = 0, r = 3, mid = 1
// mergeSort(arr, 0, 1);  l = 0, r = 1, mid = 0
// // mergeSort(arr, 0, 0);  l = 0, r = 0, mid = 0
// mergeSort(arr, 1, 1)   l = 1, r = 1, mid = 1
//
// mergeSort(arr, 0, 6);  l = 0, r = 6, mid = 3
// mergeSort(arr, 0, 3);  l = 0, r = 3, mid = 1
// mergeSort(arr, 0, 1);  l = 0, r = 1, mid = 0
// // mergeSort(arr, 0, 0);  l = 0, r = 0, mid = 0
// // mergeSort(arr, 1, 1) l = 1, r = 1, mid = 1
// merge(arr, 0, 1, 0);
//
// mergeSort(arr, 0, 6);  l = 0, r = 6, mid = 3
// mergeSort(arr, 0, 3);  l = 0, r = 3, mid = 1
// // mergeSort(arr, 0, 1);  l = 0, r = 1, mid = 0
// // mergeSort(arr, 0, 0);  l = 0, r = 0, mid = 0
// // mergeSort(arr, 1, 1) l = 1, r = 1, mid = 1
// // merge(arr, 0, 1, 0);
// mergeSort(arr, 2, 3);

File: github_code/bucket-sort/cpp/84.c
/* Implementation of bucketsort in C++ */
#include <vector>
#include <cmath>
#include <algorithm>
#include <iostream>

struct Node {
    int data;
    Node* next;
};

void split(Node*& list1, Node*& list2) {
    /* A possible implementation of split function is assigning the elements in odd position to list2 */  
    if (list1 == nullptr || list1->next == nullptr)
        return;
        
    Node* l = list1->next;
    list1->next = l->next;
    l->next = list2;
    list2 = l;
    split(list1->next, list2);
}

void merge(Node*& list1, Node* list2) {
    if (list1 == nullptr) {list1 = list2; return;}
    if (list2 == nullptr) {return;}
    if (list1->data <= list2->data) {
        merge(list1->next, list2);
    }
    else {
        merge(list2->next, list1);
        list1 = list2;
    }
}

void mergesort(Node*& list1) {
    if (list1 == nullptr || list1->next == nullptr)
        return;
        
    Node* list2 = nullptr;
    /* split list1 into two lists and assign the second half to list2 */
    split(list1, list2);
    /* order the first list */
    mergesort(list1);
    /* order the second list */
    mergesort(list2);
    /* merge the two lists */
    merge(list1, list2);
}

int msbits(int x, int minValue, int n) {
    return (int)(floor((x - minValue)/n));
}

void bucketsort(std::vector<int>& arr) {
    int n = arr.size();
    int minValue = *(std::min_element(arr.begin(), arr.end()));   
    
    /* create n empty buckets */
    Node* bucketList[n];
    
    for (int i = 0; i < n; i++) {
        bucketList[i] = nullptr;
    }
    
    /* fill buckets with elements from the original array */
    for (int i = 0; i < n; i++) {
        int index = msbits(arr[i],minValue,n); 
        Node* p = new Node;
        p->data = arr[i];
        p->next = bucketList[index];
        bucketList[index] = p;
    }
    
    /* sort individual buckets */
    for (int i = 0; i < n; i++) {
        mergesort(bucketList[i]);
    }
    
    /* concatenate all buckets */
    int k = 0;
    for (int i = 0; i < n; i++) {
        for (Node* p = bucketList[i]; p != nullptr; p = p->next) {
            arr[k++] = p->data;    
        }
    }
}

int main()
{
    std::vector<int> arr = {29,25,3,49,9,37,21,43};
    bucketsort(arr);
    for (auto i: arr) {
        std::cout << i << " ";
    }
    
    return 0;
}
